modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/util/IgniteMath.java

```diff
@@ -308,17 +297,20 @@ public static short convertToShortExact(long x) {
 
     /** Cast value to {@code short}, throwing an exception if the result overflows an {@code short}. */
     public static short convertToShortExact(double x) {
-        if (x > Short.MAX_VALUE || x < Short.MIN_VALUE)
+        if (x < (Double)SHORT_BOUNDS[2] || x > (Double)SHORT_BOUNDS[3])
             throw new ArithmeticException(SMALLINT.getName() + " overflow");
 
-        return (short)x;
+        return (short)round(x);
     }
 
     /** Cast value to {@code short}, throwing an exception if the result overflows an {@code short}. */
     public static short convertToShortExact(Number x) {
-        checkNumberLongBounds(SMALLINT, x);
```

Comment:
Perhaps we can round x, check bounds for long value, convert it to long and leave convertToShortExact? In this case we don't need checkShortBounds and other boilerplate methods.
Something like this:
```
x = round(x);
checkNumberLongBounds(SMALLINT, x);
return convertToShortExact(x.longValue());
```
Where `round`:
```
    private static Number round(Number x) {
        return convertToBigDecimal(x).setScale(0, NUMERIC_ROUNDING_MODE);
    }
```

Reply:
Attached v2 : https://github.com/apache/ignite/pull/11644